File: 0127-word-ladder/0127-word-ladder.cpp
int init=[] {ios_base::sync_with_stdio(false);cin.tie(0);return 0;}();
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_map<string, vector<string>> graph;

        for (const string& word : wordList) {
            for (size_t i = 0; i < beginWord.length(); ++i) {
                string key = word.substr(0, i) + '*' + word.substr(i + 1);
                graph[key].push_back(word);
            }
        }

        queue<pair<string, int>> queue;
        queue.push({beginWord, 1});
        unordered_set<string> visited;
        visited.insert(beginWord);

        while (!queue.empty()) {
            auto [cur, level] = queue.front();
            queue.pop();

            for (size_t i = 0; i < beginWord.length(); ++i) {
                string inter = cur.substr(0, i) + '*' + cur.substr(i + 1);

                for (const string& word : graph[inter]) {
                    if (word == endWord) {
                        return level + 1;
                    }

                    if (visited.find(word) == visited.end()) {
                        visited.insert(word);
                        queue.push({word, level + 1});
                    }
                }

                graph[inter].clear();
            }
        }

        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.rb
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}
def ladder_length(begin_word, end_word, word_list)
  graph = Hash.new { |hash, key| hash[key] = [] }

  word_list.each do |word|
    (0...begin_word.length).each do |i|
      graph[word[0...i] + '*' + word[i+1..-1]] << word
    end
  end

  queue = [[begin_word, 1]]
  visited = { begin_word => true }

  while !queue.empty?
    cur, level = queue.shift
    (0...begin_word.length).each do |i|
      inter = cur[0...i] + '*' + cur[i+1..-1]

      graph[inter].each do |word|
        if word == end_word
          return level + 1
        end

        if !visited[word]
          visited[word] = true
          queue << [word, level + 1]
        end
      end

      graph[inter] = []
    end
  end

  return 0
end